langchain/chains/constitutional_ai/base.py

```diff
@@ -0,0 +1,129 @@
+"""Chain for applying constitutional principles to the outputs of another chain."""
+from typing import Any, Dict, List
+
+from langchain.chains.base import Chain
+from langchain.chains.constitutional_ai.models import ConstitutionalPrinciple
+from langchain.chains.constitutional_ai.prompts import CRITIQUE_PROMPT, REVISION_PROMPT
+from langchain.chains.llm import LLMChain
+from langchain.llms.base import BaseLLM
+from langchain.prompts.few_shot import FewShotPromptTemplate
+
+
+class ConstitutionalChain(Chain):
+    """Chain for applying constitutional principles.
+
+    Example:
+        .. code-block:: python
+
+            from langchain import ConstitutionalChain, OpenAI, LLMChain
+
+            qa_prompt = PromptTemplate(
+                template="Q: {question} A:",
+                input_variables=["question"],
+            )
+            qa_chain = LLMChain(llm=OpenAI(), prompt=qa_prompt)
+
+            constitutional_chain = ConstitutionalChain(
+                chain=qa_chain,
+                constitutional_principles=[
+                    ConstitutionalPrinciple(
+                        critique_request="Tell if this answer is good.",
+                        revision_request="Give a better answer.",
+                    )
+                ],
+            )
+
+            constitutional_chain.run(question="What is the meaning of life?")
+    """
+
+    chain: LLMChain
+    constitutional_principles: List[ConstitutionalPrinciple]
+    critique_chain: LLMChain
+    revision_chain: LLMChain
+    llm: BaseLLM
+
+    @classmethod
+    def from_llm(
+        cls,
+        llm: BaseLLM,
+        chain: LLMChain,
+        critique_prompt: FewShotPromptTemplate = CRITIQUE_PROMPT,
+        revision_prompt: FewShotPromptTemplate = REVISION_PROMPT,
+        **kwargs: Any,
+    ) -> "ConstitutionalChain":
+        """Create a chain from an LLM."""
+        critique_chain = LLMChain(llm=llm, prompt=critique_prompt)
+        revision_chain = LLMChain(llm=llm, prompt=revision_prompt)
+        return cls(
+            chain=chain,
+            critique_chain=critique_chain,
+            revision_chain=revision_chain,
+            llm=llm,
+            **kwargs,
+        )
+
+    @property
+    def input_keys(self) -> List[str]:
+        """Defines the input keys."""
+        return self.chain.input_keys
+
+    @property
+    def output_keys(self) -> List[str]:
+        """Defines the output keys."""
+        return ["output"]
+
+    def _call(self, inputs: Dict[str, str]) -> Dict[str, str]:
+        response = self.chain.run(**inputs)
+        input_prompt = self.chain.prompt.format(**inputs)
+
+        self.callback_manager.on_text(
+            text="Initial response: " + response + "\n\n",
+            verbose=self.verbose,
+            color="yellow",
+        )
+
+        for constitutional_principle in self.constitutional_principles:
+            critique = self._parse_critique(
```

Comment:
nit: a lot of nested stuff - would be good to split out to separate lines to avoid hanging indents, eg (better naming needed)

```
res = self.critique_chain.run(...)
critiqque = self._parse_critique(res, ...)
```